### ISSUE TITLE
Writing BAM to file

### ISSUE THREAD

**Issue by martinjvickers** [seqan/seqan#2049]
**TL:DR**
I'm trying to create a BAM file from scratch and write it out. I want to be able to do this record by record.

I've been struggling with this on and off for a while and I'm clearly missing a crucial element somewhere. All of the examples of writing out to a BAM file in the how-to assume you've already read in a BAM file therefore you copy the context, e.g.

`BamFileOut bamFileOut(context(bamFileIn), std::cout, Bam());`

I was directed to the following example of creating a BAM from scratch from the io tests;

[https://github.com/seqan/seqan/blob/master/tests/bam_io/test_write_bam.h](https://github.com/seqan/seqan/blob/master/tests/bam_io/test_write_bam.h)

which is great however this writes to a `String<char>` using the `write()` function which I can't seem to get to write out to anything other than the `String<char>`. This would also mean that everything is stored `String<char>` before being written out to the BAM, therefore consuming RAM.

Ideally what I'd like to do is to be able to append records to a BAM file in a loop like the how-to examples show however I've not found a way to create a new `BamFileOut` with a newly created context. I thought something like this to maybe work;

```
StringSet<CharString> referenceNameStore;
NameStoreCache<StringSet<CharString> > referenceNameStoreCache(referenceNameStore);
BamIOContext<StringSet<CharString> > bamIOContext(referenceNameStore, referenceNameStoreCache);
BamFileOut bamFileOut(bamIOContext, toCString("out.bam"), Bam());
```

However this doesn't appear possible to use bamFileOut in this way. Does anyone have any advice/code snippets etc.?

**Comment by martinjvickers** [comment]
As an aside, I was getting some mileage out of doing something like this (psudo-code);

```
BamFileOut bamFileOut(toCString("out.bam")); 
BamHeader header = createHeader();
writeHeader(bamFileOut, header);

while(!endOfProcessing)
{
      BamAlignmentRecord currRec = createRecord();
      writeRecord(bamFileOut, currRec);
}
```

This creates a sensibly sized BAM which samtools will show a fully readable header along with the first read in my test file which happens to not align. It then dies due to it being a truncated file. 

I understand the reason for this, that I've not been able to give bamFileOut a context. I just don't know how to resolve it if this is the way it should be done.

**Comment by JakobSch** [comment]
Hi,
nice linking to this thread. The official test told me something new. In the past, I tried writing several BamHeaderRecords to a BamFileOut which created trouble when writing .bam files. Obviously one have to write only one BamHeader and append several BamHeaderRecords to it using appendValue().

However isn't this thread here answered?
The code you posted in my question seems to work.

**Comment by martinjvickers** [comment]
Hi @JakobSch. The code I posted in your question, #2068 does produce a valid SAM file but it segfaults when trying to produce a BAM file. I was hoping to get a good example of how to do it so that we can add to the manual.

**Comment by martinjvickers** [comment]
#2068 goes some way in answering some questions about how to achieve what I need (the reason I opened this issue) however I'm still unable to get it to do what I need.

The below small example demonstrates what I am trying to do (the `CharString input` data would be a stream coming from another aligner, e.g. bowtie2 which would be intercepted/processed/analysed and the result written to a BAM file.

I get a Segmentation Fault whenever I try to write out a mapped read to the file. 

```
#include <iostream>
#include <seqan/sequence.h>  // CharString, ...
#include <seqan/stream.h>    // to stream a CharString into cout
#include <seqan/file.h>
#include <seqan/arg_parse.h>
#include <seqan/seq_io.h>
#include <seqan/bam_io.h>
using namespace seqan;
using namespace std;

int main(int argc, char const ** argv)
{

        CharString input =
                "@HD    VN:1.0  SO:unsorted\n"
                "@SQ    SN:Chr1 LN:30427671\n"
                "@SQ    SN:Chr2 LN:19698289\n"
                "@SQ    SN:Chr3 LN:23459830\n"
                "@SQ    SN:Chr4 LN:18585056\n"
                "@SQ    SN:Chr5 LN:26975502\n"
                "@SQ    SN:ChrC LN:154478\n"
                "@SQ    SN:ChrM LN:366924\n"
                "HS3:420:C3EHMACXX:8:1101:1230:2142     0       Chr1    10962037        255     100M    *       0       0       AGTAAGAATTTAGAGAGTGATAGAGAATTTAATTGTGTATTATAGGTGGATAAAGTTATTGTGGTATTTTTTATGAGATTTAAAGATTTCGTAGATACGA    @@@DDFFDFH?DHIIIIB9EFACF@BFGGCFH<H9EFIH9FG>F*C1?D=?GGCFDFAGGFFCG=BBHEH);DD>@DCEHFGEECCDDEF;AC=@DDC1=    AS:i:-10        XN:i:0  XM:i:2  XO:i:0  XG:i:0  NM:i:2  MD:Z:89T7T2     YT:Z:UU\n"
                "HS3:420:C3EHMACXX:8:1101:1302:2100     4       *       0       0       *       *       0       0       ATTTAATGATTTTTGAAAGTTATTTGTTAATATTTTTTTTTTTTGGAAATTTGGTTTTTTTGTTTTATGATTTTTTGTATTTGTTTGATAAAAATTTGTA    ?@@DDDDDHBAFHHE=@E<IHECEG<,AFHCG>FA<?F@@FHBE########################################################    YT:Z:UU\n"
                "HS3:420:C3EHMACXX:8:1101:1353:2143     4       *       0       0       *       *       0       0       TTGATTGGGTATTAAAAATATTTTTTTTTTTTTTTAATATATTTTTTTTAAAAAATCAATTTTTAAACTAAAAATTGATTTTTTTTTATTTTTTTTAAAG    @@??DD>;?AD?DGHIIICHHG@GGIGIEGEA@B##################################################################    YT:Z:UU\n"
                "HS3:420:C3EHMACXX:8:1101:1272:2165     0       Chr3    12595064        255     100M    *       0       0       ATTGTTTAGTTTTTTAATTAGATTTTTGTTTTTTTTGTATTTATATAATAAATATTTTGTGAGATTGTTTAATTAATATTTATATGAATGTTAATTTGTA    @@CDDEFDD<DFHIIGGIIHGCEGIIIEDAGGIIGIIGIG@HHIIGII@EGHGIFIH:=?CE@DDFEDCECCECD@ADCDEECCD@CDD@CDD3;@DCCD    AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:100        YT:Z:UU\n";

        Iterator<CharString, Rooted>::Type iter = begin(input);

        StringSet<CharString> referenceNameStore;
        NameStoreCache<StringSet<CharString> > referenceNameStoreCache(referenceNameStore);
        BamIOContext<StringSet<CharString> > bamIOContext(referenceNameStore, referenceNameStoreCache);

        BamFileOut bamFileOut(toCString("test.sam"));
        BamHeader header;
        readHeader(header, bamIOContext, iter, Sam());
        writeHeader(bamFileOut, header);

        String<BamAlignmentRecord> alignments;
        while (!atEnd(iter))
        {
                resize(alignments, length(alignments) + 1);
                readRecord(back(alignments), bamIOContext, iter, Sam());
        }

        for(auto& i : alignments)
        {
                writeRecord(bamFileOut, i);
        }

        return 0;
}
```

@h-2 Is it possible could take a quick look at this, I first posted this issue quite a while ago.

**Comment by JakobSch** [comment]
I exchanged .sam with .bam and it worked.
Have you tried my cmake script?
Possibly your error has something to do with zlib, as it is not only needed but has also to be loaded before seqan.

**Comment by JakobSch** [comment]
Ok, now to the code from your last code here:

When I compile and run it I get the error:

    terminate called after throwing an instance of 'seqan::ParseError'
      what():  Unexpected character '\n' found.
    Aborted

just removing the last or all of them did not work, so further debugging would take much more time.

**Comment by martinjvickers** [comment]
Hi @JakobSch. Thanks for helping out and taking a look. 

That's interesting, I don't get that at all. I just get the following;
```
./write_sam_from_stream 
Segmentation fault (core dumped)
```
I've added this example to a repo (https://github.com/martinjvickers/write_sam_from_stream) and added a travis build for it just to ensure it's not my desktop here doing something odd and that does the same thing (https://travis-ci.org/martinjvickers/write_sam_from_stream/builds/235630154)

I don't think that it's the `\n` because if you keep only the header part of the stream it works;

```
$ samtools view -h test.sam 
@HD	VN:1.0	SO:unsorted
@SQ	SN:Chr1	LN:30427671
@SQ	SN:Chr2	LN:19698289
@SQ	SN:Chr3	LN:23459830
@SQ	SN:Chr4	LN:18585056
@SQ	SN:Chr5	LN:26975502
@SQ	SN:ChrC	LN:154478
@SQ	SN:ChrM	LN:366924
```
Also, if you keep the unmapped reads but remove the mapped reads it works

```
$ ./write_sam_from_stream 
$ samtools view -h test.sam 
@HD	VN:1.0	SO:unsorted
@SQ	SN:Chr1	LN:30427671
@SQ	SN:Chr2	LN:19698289
@SQ	SN:Chr3	LN:23459830
@SQ	SN:Chr4	LN:18585056
@SQ	SN:Chr5	LN:26975502
@SQ	SN:ChrC	LN:154478
@SQ	SN:ChrM	LN:366924
HS3:420:C3EHMACXX:8:1101:1302:2100	4	*	0	0	*	*	0	0	ATTTAATGATTTTTGAAAGTTATTTGTTAATATTTTTTTTTTTTGGAAATTTGGTTTTTTTGTTTTATGATTTTTTGTATTTGTTTGATAAAAATTTGTA	?@@DDDDDHBAFHHE=@E<IHECEG<,AFHCG>FA<?F@@FHBE########################################################	YT:Z:UU
HS3:420:C3EHMACXX:8:1101:1353:2143	4	*	0	0	*	*	0	0	TTGATTGGGTATTAAAAATATTTTTTTTTTTTTTTAATATATTTTTTTTAAAAAATCAATTTTTAAACTAAAAATTGATTTTTTTTTATTTTTTTTAAAG	@@??DD>;?AD?DGHIIICHHG@GGIGIEGEA@B##################################################################	YT:Z:UU
```

I think the issue is that somehow I'm not associating the chromosomes in the stream with those in the header correctly.

**Comment by martinjvickers** [comment]
@JakobSch The key to solving this was the line in your previous issue #2068 
```
bamFileOut.context = bamIOContext;
```
By adding that to mine I was able to get it all to work correctly. I've updated the working example on my [github repository](https://github.com/martinjvickers/write_sam_from_stream).

Thank you for your help, it was useful to debug out loud.

**Comment by JakobSch** [comment]
Glad I could help :)
